docs/machine-learning/Recurrent_Neural_Network.md

```diff
@@ -0,0 +1,174 @@
+
+4. **Bidirectional RNN (BiRNN)**:  
+   In a bidirectional RNN, two RNNs are run in parallel: one processes the sequence from left to right, and the other processes it from right to left. This architecture is useful for tasks where future context is as important as past context.
+
+### Problem Statement:
+Given a sequential dataset (e.g., a sentence for language modeling or a time-series for prediction), the goal of an RNN is to predict the next element in the sequence or classify the entire sequence based on its content.
+
+### Key Concepts:
+- **Sequential Processing**:  
+  RNNs process input sequences step by step, maintaining a hidden state that is updated at each step based on both the current input and previous hidden state.
+
+- **Vanishing Gradient Problem**:  
+  In vanilla RNNs, gradients may become very small (vanish) during backpropagation through time (BPTT), making it difficult to learn long-term dependencies. This is alleviated by architectures like LSTM and GRU.
+
+- **Backpropagation Through Time (BPTT)**:  
+  RNNs use BPTT to train the network, where gradients are propagated backward through each time step in the sequence.
+
+- **Long-term Dependencies**:  
+  RNNs struggle with learning long-term dependencies (i.e., when the gap between relevant inputs and outputs is large), but LSTMs and GRUs address this challenge by incorporating gating mechanisms.
+
+### Steps Involved:
+1. **Input Data**:  
+   Sequential data, such as a series of time steps (e.g., stock prices, words in a sentence), is fed into the RNN one time step at a time.
+
+2. **Hidden State Update**:  
+   At each time step, the hidden state is updated based on the current input and the previous hidden state, capturing information from earlier time steps.
+
+3. **Output Generation**:  
+   The output can be generated at each time step or at the end of the sequence, depending on the task. In tasks like language modeling, the RNN predicts the next word at each step.
+
+4. **Loss Calculation and Backpropagation**:  
+   The loss is calculated based on the predicted output and the true target. Backpropagation through time (BPTT) is used to update the weights.
+
+5. **Final Prediction**:  
+   The RNN produces a final prediction, which could be a classification of the entire sequence (e.g., sentiment analysis) or a prediction for the next time step in the sequence (e.g., time-series forecasting).
+
+### Split Criteria:
+RNNs handle sequential data where each input depends on the previous elements. The network processes the data one time step at a time, updating the hidden state with each new input.
+
+### Time Complexity:
+- **Training Complexity**:  
+  The time complexity of training an RNN depends on the number of time steps \( T \), the size of the input \( n \), and the number of hidden units \( h \). It is roughly \( O(T \cdot n \cdot h^2) \).
+
```

Comment:
replace `\( T \)` with `$T$`, `\( n \)`, with `$n$`, `\( h \)` with `$h$`, and `\( O(T \cdot n \cdot h^2) \)` with `$ O(T \cdot n \cdot h^2) $`